clippy_lints/src/if_let_mutex.rs

```diff
@@ -0,0 +1,137 @@
+use crate::utils::{match_type, paths, span_lint_and_help};
+use if_chain::if_chain;
+use rustc_hir::{Expr, ExprKind, MatchSource, StmtKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for `Mutex::lock` calls in `if let` expression
+    /// with lock calls in any of the else blocks.
+    ///
+    /// **Why is this bad?** The Mutex lock remains held for the whole
+    /// `if let ... else` block and deadlocks.
+    ///
+    /// **Known problems:** This lint does not generate an auto-applicable suggestion.
```

Comment:
```suggestion
    /// **Known problems:** None.
```

This is not really necessary here, since one might not expect a suggestion here, unlike with the `macro_use` lint